File: Sorting Searching/MergeSort.py
<<<<<<< HEAD
def merge(arr, start, end, mid):
    res = []
    j = mid+1
    i = start
    while i <= mid and j <= end:
        if(arr[i] < arr[j]):
            res.append(arr[i])
            i += 1
        else:
            res.append(arr[j])
            j += 1
    while i <= mid:
        res.append(arr[i])
        i += 1
    while j <= end:
        res.append(arr[j])
        j += 1
    for x in range(start, end+1):
        arr[x] = res[x-start]

def inplace_merge(arr, start, end, mid):
    i=start
    j=mid+1
    if arr[mid] <= arr[j]:
        return
    while mid < end:
        if(arr[i]<=arr[j]):
            i += 1
            mid += 1
        else:
            temp = arr[j]
            k = j
            while k > i:
                arr[k] = arr[k-1]
                k -= 1
            arr[i] = temp
            j += 1
            i += 1
            mid += 1

def merge_sort(arr, start, end):
    if(start < end):
        mid = (start+end) // 2
        merge_sort(arr, start, mid)
        merge_sort(arr, mid+1, end)
        inplace_merge(arr, start, end, mid)

a = [5, 4, 3, 2, 1]
merge_sort(a, 0, len(a)-1)
=======
def merge(arr, start, end, mid):
    res = []
    j = mid+1
    i = start
    while i <= mid and j <= end:
        if(arr[i] < arr[j]):
            res.append(arr[i])
            i += 1
        else:
            res.append(arr[j])
            j += 1
    while i <= mid:
        res.append(arr[i])
        i += 1
    while j <= end:
        res.append(arr[j])
        j += 1
    for x in range(start, end+1):
        arr[x] = res[x-start]

def inplace_merge(arr, start, end, mid):
    i=start
    j=mid+1
    if arr[mid] <= arr[j]:
        return
    while mid < end:
        if(arr[i]<=arr[j]):
            i += 1
            mid += 1
        else:
            temp = arr[j]
            k = j
            while k > i:
                arr[k] = arr[k-1]
                k -= 1
            arr[i] = temp
            j += 1
            i += 1
            mid += 1

def merge_sort(arr, start, end):
    if(start < end):
        mid = (start+end) // 2
        merge_sort(arr, start, mid)
        merge_sort(arr, mid+1, end)
        inplace_merge(arr, start, end, mid)

a = [5, 4, 3, 2, 1]
merge_sort(a, 0, len(a)-1)
>>>>>>> 24a80fac1064910ae51f09e81777487a1aaee3b1
print(a)

File: Sorting Searching/MergeSort.c
#include<stdio.h>
#include<stdlib.h>
#include<conio.h>

void merge(int arr[], int start, int mid, int end){
    int n1 = mid - start + 1;
    int n2 = end - mid;
    int *left = (int *)malloc(n1 * sizeof(int));
    int *right = (int *)malloc(n2 * sizeof(int));
    for(int i = 0; i < n1; i++){
        left[i] = arr[start + i];
    }
    for(int i = 0; i < n2; i++){
        right[i] = arr[mid + 1 + i];
    }
    int i = 0, j = 0, k = start;
    while(i < n1 && j < n2){
        if(left[i] <= right[j]){
            arr[k] = left[i];
            i++;
        }
        else{
            arr[k] = right[j];
            j++;
        }
        k++;
    }
    while(i < n1){
        arr[k] = left[i];
        i++;
        k++;
    }
    while(j < n2){
        arr[k] = right[j];
        j++;
        k++;
    }
    free(left);
    free(right);
}

void mergeSort(int arr[], int start, int end){
    if(start < end){
        int mid = (start + end) / 2;
        mergeSort(arr, start, mid);
        mergeSort(arr, mid + 1, end);
        merge(arr, start, mid, end);
    }
}

int main(){
    int n;
    printf("Enter the number of elements: ");
    scanf("%d", &n);
    int *arr = (int *)malloc(n * sizeof(int));
    printf("Enter the elements: ");
    for(int i = 0; i < n; i++){
        scanf("%d", &arr[i]);
    }
    mergeSort(arr, 0, n - 1);
    printf("The sorted array is: ");
    for(int i = 0; i < n; i++){
        printf("%d ", arr[i]);
    }
    free(arr);
    return 0;
}